best-practices/README.md

```diff
@@ -104,11 +104,18 @@ Rails
 Testing
 -------
 
+(See also the rspec style guide that appears in this section)
+
+* First #describe what you are doing.
+* Then establish the #context.
+* #it only expects one thing (within reason).
+* Prefer explicitness - #it #its #specify #subject are nice shortcuts but
+consider whether they are sacrificing readability.
 * Avoid `any_instance` in rspec-mocks and mocha. Prefer [dependency injection].
-* Avoid `its`, `let`, `let!`, `specify`, `before`, and `subject` in RSpec.
 * Avoid using instance variables in tests.
 * Disable real HTTP requests to external services with
-  `WebMock.disable_net_connect!`.
+`WebMock.disable_net_connect!`. Use vcr if you need to create a fixture with real-life
+requests and responses from an external service.
 * Don't test private methods.
```

Comment:
`IMO private methods are just workers for the public ones. Test the public ones correctly and you've integration tested the private methods.`

I get that. Tests on the public methods give you coverage over the private methods too, so any error in the private methods should be caught by the specs of the public methods.

That's great, but you still have to figure out what's wrong and fix it. We could test our app solely with integration tests, and we could get pretty good test coverage. But (ignoring test run times completely) debugging would be a world of pain.

There must be some ideal number of executable LOC to cover in a single test so that the test isn't excessively fine, but also isn't covering so much complexity that figuring out a spec failure requires using a debugger/lots of puts to spelunk through the code and find where the failure really is.

So, I would argue, if you feel like you want to spec a (set of) private method(s), e.g. in a well factored class with <100 LOC but only one public method, it's probably because of the number of executable LOC (regardless of which method they're in) covered by that one spec.

You could promote some private methods to public, since they're wrapping something a bit complex, but if there's no caller for those methods you're probably just complicating your API to suit an overly general rule of thumb.

You could just spec a private method that covers a bit less complexity and then stub that method in the spec for the public method, but apart from being a bit hacky it seems like in that case you'd be ignoring what your specs are telling you.

If there's that much complexity wrapped up in a public method call that's supposed to just do one thing, then you're probably violating the SRP somewhere. It seems to me there's a good chance you can pull out a bunch of logic into another class and separate out a responsibility that makes everything cleaner.